tensorflow_quantum/core/src/circuit_parser_qsim.cc

```diff
@@ -0,0 +1,373 @@
+#include "cirq/google/api/v2/program.pb.h"
+#include "tensorflow/core/lib/core/status.h"
+
+namespace tfq {
+
+using ::cirq::google::api::v2::Moment;
+using ::cirq::google::api::v2::Operation;
+using ::cirq::google::api::v2::Program;
+using ::tensorflow::Status;
+
+namespace {
+
+typedef absl::flat_hash_map<std::string, std::pair<int, float>> SymbolMap;
+typedef qsim::Cirq::GateCirq<float> QsimGate;
+typedef qsim::Circuit<QsimGate> QsimCircuit;
+
+inline Status ParseProtoArg(const Operation& op, const std::string& arg_name,
+                            const SymbolMap& param_map, float* result) {
+  // find arg_name in proto.
+  const auto arg_v = op.args().find(arg_name);
+  if (arg_v == op.args().end()) {
+    return Status(tensorflow::error::INVALID_ARGUMENT,
+                  "Could not find arg: " + arg_name + " in op.");
+  }
+  // find proto arg field.
+  const auto proto_arg = arg_v->second;
+  *result = proto_arg.arg_value().float_value();
+  if (!proto_arg.symbol().empty()) {
+    // find symbol value in param_map.
+    const auto iter = param_map.find(proto_arg.symbol());
+    if (iter == param_map.end()) {
+      return Status(
+          tensorflow::error::INVALID_ARGUMENT,
+          "Could not find symbol in parameter map: " + proto_arg.symbol());
+    }
+    *result = iter->second.second;
+  }
+  return Status::OK();
+}
+
+// series of fixed signature gate builders.
+// there is no need to error check for unparseable symbols
+// or proto args not being present. Those errors are caught
+// upstream.
+
+// single qubit gate Create(time, q0)
+inline Status SingleConstantGate(
+    const Operation& op, const SymbolMap& param_map,
+    const std::function<QsimGate(unsigned, unsigned)>& create_f,
+    const unsigned time, QsimCircuit* circuit) {
+  int q0;
+  bool unused = absl::SimpleAtoi(op.qubits(0).id(), &q0);
+  circuit->gates.push_back(create_f(time, q0));
+  return Status::OK();
+}
+
+// two qubit gate Create(time, q0, q1)
+inline Status TwoConstantGate(
+    const Operation& op, const SymbolMap& param_map,
+    const std::function<QsimGate(unsigned, unsigned, unsigned)>& create_f,
+    const unsigned time, QsimCircuit* circuit) {
+  int q0, q1;
+  bool unused = absl::SimpleAtoi(op.qubits(0).id(), &q0);
+  unused = absl::SimpleAtoi(op.qubits(1).id(), &q1);
+  circuit->gates.push_back(create_f(time, q0, q1));
+  return Status::OK();
+}
+
+// single qubit eigen -> Create(time, q0, exponent, global_shift)
+inline Status SingleEigenGate(
+    const Operation& op, const SymbolMap& param_map,
+    const std::function<QsimGate(unsigned, unsigned, float, float)>& create_f,
+    const unsigned time, QsimCircuit* circuit) {
+  int q0;
+  bool unused;
+  float exp, exp_s, gs;
+  Status u;
+  unused = absl::SimpleAtoi(op.qubits(0).id(), &q0);
+  u = ParseProtoArg(op, "exponent", param_map, &exp);
+  if (!u.ok()) {
+    return u;
+  }
+  u = ParseProtoArg(op, "exponent_scalar", param_map, &exp_s);
+  if (!u.ok()) {
+    return u;
+  }
```

Comment:
I remember that there exists a gate with `exponent` having no `exponent_scalar` or vice versa in Cirq. These lines are guaranteed in qsim?

Reply:
We actually add `exponent_scalar` in our `serializer.py` so it is gauranteed that we will have it.

Reply:
Thanks a lot!